File: csharp/protein-translation/ProteinTranslation.cs
public static class ProteinTranslation
{
    public static string[] Proteins(string strand)
    {
        Dictionary<string, string> proteinMap = new()
        {
            {"AUG", "Methionine"},
            {"UUU", "Phenylalanine"},
            {"UUC", "Phenylalanine"},
            {"UUA", "Leucine"},
            {"UUG", "Leucine"},
            {"UCU", "Serine"},
            {"UCC", "Serine"},
            {"UCA", "Serine"},
            {"UCG", "Serine"},
            {"UAU", "Tyrosine"},
            {"UAC", "Tyrosine"},
            {"UGU", "Cysteine"},
            {"UGC", "Cysteine"},
            {"UGG", "Tryptophan"},
            {"UAA", "STOP"},
            {"UAG", "STOP"},
            {"UGA", "STOP"},
        };
        List<string> proteins = new();
        int i = 0;
        while (i < strand.Length)
        {
            string condon = strand.Substring(i, 3);
            if (proteinMap[condon] == "STOP")
            {
                break;
            }
            proteins.Add(proteinMap[condon]);
            i += 3;
        }
        return proteins.ToArray();
    }
}

File: typescript/protein-translation/protein-translation.ts
type ProteinMap = { [key: string]: string}
const CODON_MAP: ProteinMap = {
  AUG: `Methionine`,
  UUU: `Phenylalanine`, UUC: `Phenylalanine`,
  UUA: `Leucine`, UUG: `Leucine`,
  UCU: `Serine`, UCC: `Serine`, UCA: `Serine`, UCG: `Serine`,
  UAU: `Tyrosine`, UAC: `Tyrosine`,
  UGU: `Cysteine`, UGC: `Cysteine`,
  UGG: `Tryptophan`,
  UAA: `STOP`, UAG: `STOP`, UGA: `STOP`,
}

export function translate(rna: string): string[] {
  const proteins: string[] = [];
  const codons = (rna.match(/.{1,3}/g) || []) as string[];
  for (const codon of codons){
    const protein = CODON_MAP[codon];
    if (!protein) throw new Error(`Invalid codon`);
    if (protein === `STOP`) break;
    proteins.push(protein);
  }
  return proteins;
}
